Spigot-Server-Patches/0246-Port-Sponge-Async-lighting-patch-to-Paper.patch

```diff
@@ -0,0 +1,1221 @@
+     private long w;
+     private int x;
+-    private final ConcurrentLinkedQueue<BlockPosition> y;
++    private final ConcurrentLinkedQueue<BlockPosition> y; public ConcurrentLinkedQueue<BlockPosition> getTileEntityPosQueue() { return y; } // Paper - OBFHELPER
+     public boolean d; public void setShouldUnload(boolean unload) { this.d = unload; } public boolean isUnloading() { return d; } // Paper - OBFHELPER
+     protected gnu.trove.map.hash.TObjectIntHashMap<Class> entityCount = new gnu.trove.map.hash.TObjectIntHashMap<Class>(); // Spigot
+ 
+@@ -123,6 +123,7 @@ public class Chunk {
+     // CraftBukkit end
+ 
+     public Chunk(World world, int i, int j) {
++        lightExecutorService = world.getLightingExecutor(); // Paper - asynclighting
+         this.sections = new ChunkSection[16];
+         this.g = new byte[256];
+         this.h = new int[256];
+@@ -151,6 +152,18 @@ public class Chunk {
+     public boolean mustSave;
+     // CraftBukkit end
+ 
++    // Paper start - asynclighting
++    // Keeps track of block positions in this chunk currently queued for sky light update
++    private java.util.concurrent.CopyOnWriteArrayList<Short> queuedSkyLightingUpdates = new java.util.concurrent.CopyOnWriteArrayList<>();
++    // Keeps track of block positions in this chunk currently queued for block light update
++    private java.util.concurrent.CopyOnWriteArrayList<Short> queuedBlockLightingUpdates = new java.util.concurrent.CopyOnWriteArrayList<>();
++    private java.util.concurrent.atomic.AtomicInteger pendingLightUpdates = new java.util.concurrent.atomic.AtomicInteger();
++    private long lightUpdateTime;
++    private java.util.concurrent.ExecutorService lightExecutorService;
++    private static final List<Chunk> EMPTY_LIST = new java.util.ArrayList<>();
++    private static final BlockPosition DUMMY_POS = new BlockPosition(0, 0, 0);
++    // Paper end
++
+     public Chunk(World world, ChunkSnapshot chunksnapshot, int i, int j) {
+         this(world, i, j);
+         boolean flag = true;
+@@ -176,6 +189,7 @@ public class Chunk {
+ 
+     }
+ 
++    public boolean isAtLocation(int i, int j) { return a(i, j); } // Paper - OBFHELPER
+     public boolean a(int i, int j) {
+         return i == this.locX && j == this.locZ;
+     }
+@@ -184,6 +198,7 @@ public class Chunk {
+         return this.b(blockposition.getX() & 15, blockposition.getZ() & 15);
+     }
+ 
++    public int getHeightValue(int i, int j) { return b(i, j); } // Paper - OBFHELPER
+     public int b(int i, int j) {
+         return this.heightMap[j << 4 | i];
+     }
+@@ -199,6 +214,7 @@ public class Chunk {
+         return null;
+     }
+ 
++    public int getTopFilledSegment() { return g(); } // Paper - OBFHELPER
+     public int g() {
+         ChunkSection chunksection = this.y();
+ 
+@@ -274,6 +290,9 @@ public class Chunk {
+     }
+ 
+     private void h(boolean flag) {
++        // Paper start - asynclighting
```

Comment:
Is it intended that any calls to the previous recheckGaps is now no longer going to do anything....?